pkg/proxy/ipvs/proxier.go

```diff
@@ -808,6 +808,45 @@ func (proxier *Proxier) syncProxyRules() {
 	// activeBindAddrs represents ip address successfully bind to DefaultDummyDevice in this round of sync
 	activeBindAddrs := map[string]bool{}
 
+	hasNodePort := false
+	for _, svc := range proxier.serviceMap {
+		svcInfo, ok := svc.(*serviceInfo)
+		if ok && svcInfo.NodePort() != 0 {
+			hasNodePort = true
+			break
+		}
+	}
+
+	// Both nodeAddresses and nodeIPs can be reused for all nodePort services
+	// and only need to be computed if we have at least one nodePort service.
+	var (
+		// List of node addresses to listen on if a nodePort is set.
+		nodeAddresses []string
+		// List of node IP addresses to be used as IPVS services if nodePort is
+		// set.
+		nodeIPs []net.IP
+	)
+
+	if hasNodePort {
+		nodeAddrSet, err := utilproxy.GetNodeAddresses(proxier.nodePortAddresses, proxier.networkInterfacer)
+		if err != nil {
+			klog.Errorf("Failed to get node ip address matching nodeport cidr: %v", err)
+		}
+		if nodeAddrSet.Len() > 0 {
+			nodeAddresses = nodeAddrSet.List()
+			for _, address := range nodeAddresses {
+				if utilproxy.IsZeroCIDR(address) {
```

Comment:
Is there a case where `nodeAddresses` could contain 0.0.0.0/0 and other addresses?
Because we could
- if `nodeAddrSet.Len()==1` and nodeAddrSet[0]=`0.0.0.0/0` => iterate over `nodeAddrSet`
- otherwise iterate over `proxier.ipGetter.NodeIPs()`

Reply:
Assuming the node has three interfaces:
```
eth0:    10.0.0.1, fe80::1c00:a8ff:fe00:481f
lo0:     127.0.0.1, ::1
docker0: 172.17.0.1
```

It's possible for nodeAddresses to contain both `::/0` and another IP, but
only when mixing IPv4 and IPv6 arguments due to checks here:

https://github.com/kubernetes/kubernetes/blob/289430a83ebf9c637b9782492aa038e029ee80c5/pkg/proxy/util/utils.go#L179-L184

Starting kube-proxy with `--nodeport-addresses=::/0,10.0.0.1/32` would cause these values:

```
nodeAddresses == ["10.0.0.1", "::/0"]                    # This is the order sets.String.List would return them.
nodeIPs       == ["10.0.0.1", "127.0.0.1", "172.17.0.1]  # IPv6 address possibly here also, if there are routes using them as source.
```

Down the line, when configuring a node port service we would have:

```
lps == []utilproxy.LocalPort{
    {
        IP: "10.0.0.1",
        ...
    },
    {
        IP: "",
    },
}
```

I'm not sure this value for `lps` makes much sense but I didn't touch this part of the code. If it's okay for it to have a single empty IP entry then I guess `utilproxy.GetNodeAddresses` could be modified to return only `::/0` and them skip iterating on `nodeAddresses` if `utilproxy.IsZeroCIDR(nodeAddresses[0])` is true.